File: rs/src/dimensions.rs
use rand::Rng;

pub type LogicalCoordinate = (isize, isize);
pub type ScreenCoordinates = (usize, usize);

#[derive(Debug, Clone)]
pub struct Dimensions {
    xmin: isize,
    xmax: isize,
    ymin: isize,
    ymax: isize,
    abs: ScreenCoordinates,
    scale: isize,
}
impl Dimensions {
    pub fn new((x, y): ScreenCoordinates, scale: usize) -> Self {
        let (dx, dy) = ((x / 2) as isize, (y / 2) as isize);

        Self {
            xmin: -dx,
            xmax: dx,
            ymin: -dy,
            ymax: dy,
            abs: (x, y),
            scale: scale as isize,
        }
    }
    pub fn bounds(&self) -> (LogicalCoordinate, LogicalCoordinate) {
        ((self.xmin, self.xmax), (self.ymin, self.ymax))
    }

    pub fn out_of_bounds(&self, (x, y): LogicalCoordinate) -> bool {
        x < self.xmin*2/self.scale || x > self.xmax*2/self.scale || y < self.ymin*2/self.scale || y > self.ymax*2/self.scale
    }

    pub fn to_screen(&self, (x, y): LogicalCoordinate) -> ScreenCoordinates {
        /// if x is less than 0, return 0
        fn saturating_cast(x: isize) -> usize {
            if x < 0 { 0 } else { x as usize }
        }
        // ((x + self.xmax) as usize, (self.ymax - y) as usize)
        // if x' is negative it causes the usize cast to wrap around
        (
            saturating_cast(x + self.xmax),
            saturating_cast(self.ymax - y),
        )
    }

    pub fn to_logical(&self, (x, y): ScreenCoordinates) -> LogicalCoordinate {
        ((x as isize - self.xmax), (y as isize - self.ymax))
    }
    /// returns a random coordinate
    pub fn random(&self) -> LogicalCoordinate {
        let mut rng = rand::rng();
        (
            rng.random_range((self.xmin*2/ self.scale) as i32..=(self.xmax*2/ self.scale) as i32)
                as isize,
            rng.random_range((self.ymin*2/ self.scale) as i32..=(self.ymax*2/ self.scale) as i32)
                as isize,
        )
    }
    pub fn get_screen_bounds(&self) -> ScreenCoordinates {
        self.abs
    }
}

File: java/app/src/main/java/snakegame/Dimensions.java

package snakegame;
import snakegame.logic.Shared.Coordinate;

public class Dimensions {
         
    // new , out_of_bounds, to_screen

    public int xmin;
    public int xmax;
    public int ymin;
    public int ymax;
    public Coordinate bounds;
    int scale;
    
    public Dimensions(Coordinate bounds, int scale) {
        int dx = bounds.x()/scale;
        int dy = bounds.y()/scale;

        this.xmin = -dx;
        this.xmax = dx;
        this.ymin = -dy;
        this.ymax = dy;
        this.bounds = bounds;
        this.scale = scale;
    }

    public boolean outOfBounds(Coordinate coords) {
        // return (coords.x() < xmax && coords.x() > xmin && coords.y() < ymax && coords.y() > ymin);
        var x = coords.x();
        var y = coords.y();
        // System.out.println("xmin:" + xmin + "\n xmax:"+ xmax + "\n ymax:"+ ymin + "\n ymax:"+ ymax + "\n bounds:" + bounds);
        return x < this.xmin || x > this.xmax || y < this.ymin || y > this.ymax;
    }

    public Coordinate toScreen(Coordinate coords) {
        int x;
        int y;
        
        if ((coords.x()+bounds.x()/2) < 0) x = 0; else x = (coords.x()+bounds.x()/2);
        if ((bounds.x()/2-coords.y()) < 0) y = 0; else y = (bounds.x()/2-coords.y());

        // var sc = Coordinate.of(coords.x()+bounds.x()/2,coords.y()+bounds.y()/2);

        var sc= Coordinate.of(x, y);
        // System.err.printf("translated %s -> %s\n",coords,sc);
        return sc;
    }

    public Coordinate random() {
        // return Coordinate.of(
        // (int) (Math.random()*((xmax-xmin)/scale + 1)+xmax),
        // (int) (Math.random()*((ymax-ymin)/scale + 1)+ymin)
        // );
        return Coordinate.of(
            (int)(Math.random()*(xmax-xmin)+xmin)+1,
            (int)(Math.random()*(ymax-ymin)+ymin)+1
        );
    }
    public Coordinate getScreenDimensions(){
        return this.bounds;
    }
}